astropy/utils/tests/test_misc.py

```diff
@@ -86,6 +87,45 @@ def test_JsonCustomEncoder():
     assert newd == tmpd
 
 
+def test_JsonCustomEncoder_FITS_rec_columns():
+    columns_descr = [
+        ('B', [1, 2], np.uint8, None),
+        ('2L', [[True, False], [False, True]], np.bool_, None),
+        ('2B', [[1, 3], [11, 12]], np.uint8, None),
+        ('2I', [[-1, 3], [11, -12]], np.int16, None),
+        ('2K', [[(1 << 63)-10, 3], [11, (1 << 63) - 100]], np.int64, None),
+        ('30A', ['one', 'two'], np.string_, None),
+        ('C', [complex(1, 2), complex(3, 1)], np.string_, [[1.0, 2.0], [3.0, 1.0]]),
+        ('PI()', [[1, 2, 3], [11, 12]], np.object_, None),
+    ]
+
+    hdu = fits.BinTableHDU.from_columns([
+        fits.Column(
+                name='col_' + format.lower(), format=format,
+                array=np.array(data, dtype=dtype))
+        for i, (format, data, dtype, preform_data) in enumerate(columns_descr)
+    ])
+
+    def get_expected_value(c):
+        # we expect regular json representations of python standard objects
+        # derived from special types
+        if c[3] is not None:
+            # for peculiar types, use special python representations
+            return c[3]
+        else:
+            # for most types, input of the column is also
+            # a valid python representation of of the column
+            return c[1]
+
+    for column_descr, hdu_column in zip(columns_descr, hdu.columns):
+        assert json.dumps(hdu_column.array, cls=misc.JsonCustomEncoder) == \
+               json.dumps(get_expected_value(column_descr))
+
+    assert json.dumps(hdu.data, cls=misc.JsonCustomEncoder) == json.dumps(
+            list(zip(*[get_expected_value(c) for c in columns_descr]))
+        )
```

Comment:
This seems a bit complicated. My suggestion would be to use some test files we already have, e.g. these two ones:
```
❯ showtable astropy/io/fits/tests/data/btable.fits 
order    name     mag   Sp 
----- ---------- ----- ----
    1     Sirius -1.45  A1V
    2    Canopus -0.73 F0Ib
    3 Rigil Kent  -0.1  G2V

❯ showtable astropy/io/fits/tests/data/variable_length_table.fits 
   var     xyz [2]
---------- -------
   [45 56] 11 .. 3
[11 12 13] 12 .. 4
```
Just reading them and comparing the serialized string to a reference one.

Reply:
Since this is tested with `tolist` in `io.fits`, no need to repeat the duplicate test here.

Reply:
> Since this is tested with `tolist` in `io.fits`, no need to repeat the duplicate test here.

I made two tests for `tolist` in `io.fits` [here](https://github.com/oda-hub/astropy/blob/7edebe69b32a0b3b2491f271d0e79b9975926e23/astropy/io/fits/tests/test_table.py#L2954-L2968).

But as I understood from discussion [above](https://github.com/astropy/astropy/pull/11957#discussion_r680117162) it may be useful to have an test of jsonification independent of particular implementation.
Or should there tests for both file-read and constructed cases with tolist but only file for json?

Reply:
For json testing only with files is fine. What we want to test here is that different types of columns are correctly serialized, it doesn't matter how the table is constructed since this is tested in the `io.fits` test you added.

Reply:
I made the tests for json dumping of files here https://github.com/astropy/astropy/pull/11957/commits/c2b36a2577d0e501df1270bf7301d73b90ff9ed7

If you insist, I remove the `test_JsonCustomEncoder_FITS_rec_columns` [there](https://github.com/oda-hub/astropy/blob/c2b36a2577d0e501df1270bf7301d73b90ff9ed7/astropy/utils/tests/test_misc.py#L90), no problem.

Reply:
> For json testing only with files is fine. What we want to test here is that different types of columns are correctly serialized, it doesn't matter how the table is constructed since this is tested in the `io.fits` test you added.

Sorry did not see this before  my last comment.
It seems to me that the table is actually somehow different depending on whether it is read or constructed. What differs is how what is in the array. If `tolist` is now outputs the same - it does not matter. But, this aspect of implementation of the custom encoder may change in the future. I am curious to look a bit more what is the nature of this difference, but I could just delete the longer non-file test, no problem.

Reply:
With the latest implementation of `tolist` the behavior is the same, and the json serializer the `io.fits` internal doesn't matter. So thanks for having removed the other test, looks good now.

Reply:
It could happen in the future that json serializer will use something else than tolist, and the test will not break when expected.  Presumably, at that time a test will be added.